modules/tetramesh/glsl/tetramesh_traversal.frag

```diff
@@ -0,0 +1,374 @@
+
+    float bgDepthScreen = invalidDepth;
+
+#if defined(BACKGROUND_AVAILABLE)
+    vec2 screenTexCoords = gl_FragCoord.xy * backgroundParameters.reciprocalDimensions;
+    vec4 background = texture(backgroundColor, screenTexCoords); 
+    // prepare background color for pre-multiplied alpha blending
+    background.rgb *= background.a;
+
+    bgDepthScreen = texture(backgroundDepth, screenTexCoords).r;
+    // background depth in Data space
+    const float tBackground = convertScreenPosToDataDepth(vec3(screenTexCoords, bgDepthScreen));
+#endif // BACKGROUND_AVAILABLE
+
+    int tetraFaceId = in_frag.tetraFaceId;
+    vec3 pos = in_frag.position;
+
+    int tetraId = tetraFaceId / 4;
+    int localFaceId = tetraFaceId % 4;
+    ivec4 vertices = vertexIds[tetraId];
+
+    // determine scalar value at entry position
+    Tetra tetra = getTetra(tetraId);
+    float prevScalar = normalizeScalar(barycentricInterpolation(pos, tetra));
+
+    vec4 dvrColor = vec4(0);
+#if defined(BACKGROUND_AVAILABLE)
+    // blend background if it lies in front of the start position of the ray
+    dvrColor = tBackground < tEntry ? background : dvrColor;
+
+    // suppress background blending tests if the background alpha is zero
+    bool bgBlended = tBackground < tEntry || background.a == 0.0;
+#endif // BACKGROUND_AVAILABLE
+
+    float tTotal = tEntry;
+    float tFirstHit = invalidDepth;
+    int steps = 0;
+    while (tetraFaceId > -1 && steps < maxSteps && dvrColor.a < ERT_THRESHOLD) {
+        // find next tetra
+        tetraId = tetraFaceId / 4;
+        localFaceId = tetraFaceId % 4;
+        vertices = vertexIds[tetraId];
+
+        // query data of current tetrahedron
+        tetra = getTetra(tetraId);
+        mat4x3 faceNormal = getFaceNormals(tetra);
+
+        // intersect ray at current position with all tetra faces
+        vec4 vdir = vec4(dot(faceNormal[0], rayDirection),
+                         dot(faceNormal[1], rayDirection),
+                         dot(faceNormal[2], rayDirection),
+                         dot(faceNormal[3], rayDirection));
+        vec4 vt = vec4(dot(tetra.v[1] - pos, faceNormal[0]),
+                       dot(tetra.v[2] - pos, faceNormal[1]),
+                       dot(tetra.v[3] - pos, faceNormal[2]),
+                       dot(tetra.v[0] - pos, faceNormal[3])) / vdir;
+
+        // only consider intersections on the inside of the current triangle faces, that is t > 0.
+        // Also ignore intersections being parallel to a face
+        vt = mix(vt, vec4(invalidDepth), lessThan(vdir, vec4(0.0)));
+
+        // ignore self-intersection with current face ID, set distance to max
+        vt[localFaceId] = invalidDepth;
+
+        // closest intersection
+        // face ID of closest intersection
+        const int vface1 = vt.x < vt.y ? 0 : 1;
+        const int vface2 = vt.z < vt.w ? 2 : 3;
+        const int vface = vt[vface1] < vt[vface2] ? vface1 : vface2;        
+        const float tmin = vt[vface];
```

Comment:
If tmin is the end of the tetra it does not make sense to name it tmin?

Reply:
tmin is the minimal distance to each of the planes of the tetra.

Reply:
I guess it does not matter too much, but `tHit` could be used in that case

Reply:
tHit does not make sense since there can be multiple hits.